pkg/api/v2/types.go

```diff
@@ -0,0 +1,331 @@
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE
+
+package v2
+
+import (
+	"fmt"
+
+	"github.com/microsoft/hivedscheduler/pkg/api"
+	core "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+// PodSchedulingSpec represents HiveD scheudling spec in k8s pod request.
+type PodSchedulingSpec struct {
+	Version              string                 `yaml:"version"`              // version of HiveD PodSchedulingSpec, currently supports v1, v2.
+	VirtualCluster       api.VirtualClusterName `yaml:"virtualCluster"`       // virtual cluster for pod to be scheduled in
+	Priority             int32                  `yaml:"priority"`             // pod priority
+	PinnedCellId         api.PinnedCellId       `yaml:"pinnedCellId"`         // pinned cell id to be scheduled
+	CellType             string                 `yaml:"cellType"`             // cell type to be used in pod, can be leaf or non-leaf cell defined in config, no higher than node level
+	CellNumber           int32                  `yaml:"cellNumber"`           // cell number to be used in pod, cannot exceed node resource limit
+	GangReleaseEnable    bool                   `yaml:"gangReleaseEnable"`    // whether release in gang (all pods released at the same time) or not
+	LazyPreemptionEnable bool                   `yaml:"lazyPreemptionEnable"` // whether lazy preempt or not
+	PodRootGroup         *PodGroupSpec          `yaml:"podRootGroup"`         // the hierarchical structure for whole pod group
+}
+
+// PodGroupSpec represents a tree stucture of pod group spec.
+type PodGroupSpec struct {
+	Name          string               `yaml:"name"`          // pod group name
+	WithinOneCell api.CellType         `yaml:"withinOneCell"` // within cell for all cells in current group, e.g., two GPU-cell within one numa-cell, two node-cell within one rack-cell
+	Pods          []PodGroupMemberSpec `yaml:"pods"`          // pod list for current group, any two pods are gang with each other in the groupS
+	ChildGroups   []*PodGroupSpec      `yaml:"childGroups"`   // child group in the hierarchical structure
+}
```

Comment:
2 ChildGroups listed in a group is also always gang with each other, and the Pods?

Reply:
updated